source_middleware.go

```diff
@@ -409,3 +409,139 @@ func (s *sourceWithSchemaExtraction) encodeWithSchema(sch schema.Schema, data an
+			Default: strconv.FormatBool(*c.UseContext),
+			Description: "Specifies whether to use a schema context name. If set to false, no schema context name will " +
+				"be used, and schemas will be saved with the subject name specified in the connector " +
+				"(not safe because of name conflicts).",
+			Type: config.ParameterTypeBool,
+		},
+		c.ContextNameParameterName(): config.Parameter{
+			Default: func() string {
+				if c.ContextName == nil {
+					return ""
+				}
+
+				return *c.ContextName
+			}(),
+			Description: func() string {
+				d := "Schema context name to be used. Used as a prefix for all schema subject names."
+				if c.ContextName == nil {
+					d += " Defaults to the connector ID."
+				}
+				return d
+			}(),
+			Type: config.ParameterTypeString,
+		},
+	}
+}
+
+func (c SourceWithSchemaContextConfig) UseContextParameterName() string {
+	return "sdk.schema.context.use"
+}
+
+func (c SourceWithSchemaContextConfig) ContextNameParameterName() string {
+	return "sdk.schema.context.name"
+}
+
+// SourceWithSchemaContext is a middleware that makes it possible to configure
+// the schema context for records read by a source.
+type SourceWithSchemaContext struct {
+	Config SourceWithSchemaContextConfig
+}
+
+// Wrap a Source into the schema middleware. It will apply default configuration
+// values if they are not explicitly set.
+func (s *SourceWithSchemaContext) Wrap(impl Source) Source {
+	if s.Config.UseContext == nil {
+		t := true
+		s.Config.UseContext = &t
+	}
+
+	return &sourceWithSchemaContext{
+		Source: impl,
+		mwCfg:  s.Config,
+	}
+}
+
+type sourceWithSchemaContext struct {
+	Source
+	// mwCfg is the default middleware config
+	mwCfg SourceWithSchemaContextConfig
+
+	useContext  bool
+	contextName string
+}
+
+func (s *sourceWithSchemaContext) Parameters() config.Parameters {
+	return mergeParameters(s.Source.Parameters(), s.mwCfg.parameters())
+}
+
+func (s *sourceWithSchemaContext) Configure(ctx context.Context, cfg config.Config) error {
+	s.useContext = *s.mwCfg.UseContext
+	if useStr, ok := cfg[s.mwCfg.UseContextParameterName()]; ok {
+		use, err := strconv.ParseBool(useStr)
+		if err != nil {
+			return fmt.Errorf("could not parse `sdk.schema.context.use`, input %v: %w", useStr, err)
+		}
+		s.useContext = use
+	}
+
+	if s.useContext {
+		// the order of precedence (least to most) is:
+		// 1. connector ID (if no context name is configured anywhere)
+		// 2. default middleware config
+		// 3. user config
+		s.contextName = internal.ConnectorIDFromContext(ctx)
+		if s.mwCfg.ContextName != nil {
+			s.contextName = *s.mwCfg.ContextName
+		}
+		if ctxName, ok := cfg[s.mwCfg.ContextNameParameterName()]; ok {
+			s.contextName = ctxName
+		}
+	}
+
+	return s.Source.Configure(sdkschema.WithSchemaContextName(ctx, s.contextName), cfg)
+}
+
+func (s *sourceWithSchemaContext) Open(ctx context.Context, pos opencdc.Position) error {
+	return s.Source.Open(sdkschema.WithSchemaContextName(ctx, s.contextName), pos)
```

Comment:
Sounds good to me :+1:

Reply:
Btw, you're suggesting moving that function into `sourceWithSchemaContext`. The reason I made the function exported is because of `schemaContextNameKey`. 

The function `GetSchemaContextName` needs to access `schemaContextNameKey`, and it seemed better to me to have the key as an unexported type, then to have this method in `sourceWithSchemaContext`.

Reply:
No, no need to move anything, I'd just add this:
```
func (s * sourceWithSchemaContext) withSchemaContextName(ctx context.Context) context.Context {
	if s.contextName != "" {
		ctx = sdkschema.WithSchemaContextName(ctx, s.contextName)
	}
	return ctx
}
```

Reply:
Gotcha, but is there a specific reason for having that check in `sdkschema.WithSchemaContextName`?

Reply:
I assume you mean a reason for _**not**_ having that check in `sdkschema.WithSchemaContextName`? I think that function shouldn't be too smart and assume what the caller wants to do. If they want to store an empty string, so be it. Here we are the caller, we know what we want and what we _don't_ want - an empty string 😄

Reply:
Yup, that's what I meant, having the check in `sdkschema.WithSchemaContextName`. I'm thinking why would we actually need the check?

Reply:
Just seems cleaner if we don't attach any value if we don't want any schema context. Maybe I'm overthinking this 🤷